include/ginkgo/core/base/lin_op.hpp

```diff
@@ -855,49 +855,75 @@ public:                                                                      \
                   "semi-colon warnings")
 
 
-#if !(defined(__CUDACC__) || defined(__HIPCC__))
 /**
  * Creates a factory parameter in the factory parameters structure.
  *
  * @param _name  name of the parameter
+ * @param _type  type of the parameter
  * @param __VA_ARGS__  default value of the parameter
  *
  * @see GKO_ENABLE_LIN_OP_FACTORY for more details, and usage example
  *
  * @ingroup LinOp
  */
-#define GKO_FACTORY_PARAMETER(_name, ...)                                    \
-    mutable _name{__VA_ARGS__};                                              \
+#define GKO_FACTORY_PARAMETER(_name, _type, ...)                             \
+    _type mutable _name{__VA_ARGS__};                                        \
```

Comment:
This macro change definitively breaks the interface since it is defined in a public header and everyone who wants to add their own solver would use it. With this change, it would no longer compile.

Reply:
I have to agree with Thomas here, this is a definite interface break. I would rather that we go with a less potent version as suggested previously, so that we still fix this issue (for a set number of template parameters on CUDA) while not breaking interface.